crates/in_stream/src/mem.rs

```diff
@@ -0,0 +1,433 @@
+            Some(send) => {
+                // if we're still connected, send data to our pair
+                match send.send(buf.to_vec()) {
+                    Ok(_) => Ok(buf.len()),
+                    Err(_) => Err(ErrorKind::NotConnected.into()),
+                }
+            }
+        }
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        if self.send.is_none() {
+            return Err(ErrorKind::NotConnected.into());
+        }
+        Ok(())
+    }
+}
+
+impl InStreamStd for InStreamMem {}
+
+impl InStreamMem {
+    /// private constructor, you probably want `connect`
+    fn priv_new(
+        url: Url2,
+        send: crossbeam_channel::Sender<Vec<u8>>,
+        recv: crossbeam_channel::Receiver<Vec<u8>>,
+    ) -> Self {
+        Self {
+            url,
+            send: Some(send),
+            recv: Some(recv),
+            recv_buf: Vec::new(),
+        }
+    }
+
+    /// see std::net::TcpStream::shutdown
+    pub fn shutdown(&mut self, how: std::net::Shutdown) -> Result<()> {
+        match how {
+            std::net::Shutdown::Read => {
+                self.recv.take();
+            }
+            std::net::Shutdown::Write => {
+                self.send.take();
+            }
+            std::net::Shutdown::Both => {
+                self.recv.take();
+                self.send.take();
+            }
+        }
+        Ok(())
+    }
+}
+
+// -- utility functions -- //
+
+pub mod in_stream_mem {
+    use super::*;
+
+    /// create a unique url for binding an InStreamListenerMem instance
+    pub fn random_url(prefix: &str) -> Url2 {
+        Url2::parse(&format!(
+            "{}://{}-{}",
+            SCHEME,
+            prefix,
+            nanoid::simple().replace("_", "-").replace("~", "+"),
+        ))
+    }
+}
+
+use in_stream_mem::random_url;
+
+/// private stream pair constructor, these streams can message each other
+fn create_mem_stream_pair(url_a: Url2, url_b: Url2) -> (InStreamMem, InStreamMem) {
+    let (send1, recv1) = crossbeam_channel::unbounded();
+    let (send2, recv2) = crossbeam_channel::unbounded();
+    (
+        InStreamMem::priv_new(url_a, send1, recv2),
+        InStreamMem::priv_new(url_b, send2, recv1),
+    )
+}
+// -- singleton memory manager -- //
+
+/// private singleton for managing virtual memory listening interfaces
+struct MemManager {
+    listeners: HashMap<Url2, crossbeam_channel::Sender<InStreamMem>>,
+}
+
+impl MemManager {
+    /// create a new singleton
+    fn new() -> Self {
+        Self {
+            listeners: HashMap::new(),
+        }
+    }
+
+    /// manage binding a new MemListener interface
+    fn bind(&mut self, url: &Url2) -> Result<InStreamListenerMem> {
+        if SCHEME != url.scheme() {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                format!("mem bind: url scheme must be '{}'", SCHEME),
+            ));
+        }
+        match url.port() {
+            None | Some(0) | Some(PORT) => (),
```

Comment:
You can match on static variable patterns? Awesome- even ocaml can't do that.

Reply:
:v: